### PR TITLE
added using a ToolTip to showing much more detailed information about…

### PR DESCRIPTION
**Pull request by immeraufdemhund**
… a test run. Closes issue #75

### CODE REVIEW

File: src/nunit-gui/Model/ResultSummary.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        /// Returns the number of failed test cases (including errors and invalid tests)
+        /// </summary>
+        public int FailedCount => FailureCount + InvalidCount + ErrorCount;
+
+        /// <summary>
+        /// Returns the sum of skipped test cases, including ignored and explicit tests
+        /// </summary>
+        public int TotalSkipCount => SkipCount + IgnoreCount + ExplicitCount;
+
+        public string OverallResult { get; set; }
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: Please insert XML documentation.

File: src/nunit-gui/Model/ResultSummary.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+        public double Duration { get; set; }
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: Please insert XML documentation.

File: src/nunit-gui/Model/ResultSummary.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        public DateTime StartTime { get; set; }
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: Please insert XML documentation.

File: src/nunit-gui/Model/ResultSummary.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        public DateTime EndTime { get; set; }
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: Please insert XML documentation.

File: src/nunit-gui/Model/ResultSummary.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Gets the number of test cases for which results
+        /// have been summarized. Any tests excluded by use of
+        /// Category or Explicit attributes are not counted.
+        /// </summary>
+        public int TestCount { get; set; }
+
+        public int WarningCount { get; set; }
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: Please insert XML documentation.

File: src/nunit-gui/Model/ResultSummary.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Gets the count of inconclusive tests
+        /// </summary>
+        public int InconclusiveCount { get; set; }
+
+        /// <summary>
+        /// Returns the number of test cases that were not runnable
+        /// due to errors in the signature of the class or method.
+        /// Such tests are also counted as Errors.
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: I would think that "counted as Errors" => "counted as Failed". Otherwise we would count the testcase twice in `FailedCount`.

File: src/nunit-gui/Model/ResultSummaryCreator.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            string type = node.GetAttribute("type");
+            string status = node.GetAttribute("result");
+            string label = node.GetAttribute("label");
+
+            switch (node.Name)
+            {
+                case "test-case":
+                    summary.TestCount++;
+
+                    switch (status)
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryCreator.cs]: Perhaps extract this switch to a helper method, so that it is easier to overlook the outer switch

File: src/nunit-gui/Model/ResultSummaryReporter.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+using System.Text;
+
+namespace NUnit.Gui.Model
+{
+    public static class ResultSummaryReporter
+    {
+        public static string WriteSummaryReport(ResultSummary summary)
+        {
+            var sb = new StringBuilder();
+            using (var writer = new StringWriter(sb))
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: Why the need for the `StringWriter`? It seems that we can do the same with a `StringBuilder`.

File: src/nunit-gui/Model/ResultSummaryReporter.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            {
+                writer.WriteLabelLine("Overall result: ", summary.OverallResult);
+
+                writer.WriteSummaryCount("  Test Count: ", summary.TestCount);
+                writer.WriteSummaryCount(", Passed: ", summary.PassCount);
+                writer.WriteSummaryCount(", Failed: ", summary.FailedCount/*, ColorStyle.Failure*/);
+                writer.WriteSummaryCount(", Warnings: ", summary.WarningCount/*, ColorStyle.Warning*/);
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: Perhaps we should remove the commented code `ColorStyle....` which is not used.

File: src/nunit-gui/Model/ResultSummaryReporter.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                {
+                    writer.WriteSummaryCount("    Skipped Tests - Ignored: ", summary.IgnoreCount);
+                    writer.WriteSummaryCount(", Explicit: ", summary.ExplicitCount);
+                    writer.WriteSummaryCount(", Other: ", summary.SkipCount);
+                    writer.WriteLine();
+                }
+
+                writer.WriteLabelLine("  Start time: ", summary.StartTime.ToString("u"));
+                writer.WriteLabelLine("    End time: ", summary.EndTime.ToString("u"));
+                writer.WriteLabelLine("    Duration: ", string.Format(NumberFormatInfo.InvariantInfo, "{0:0.000} seconds", summary.Duration));
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: This does not align as pretty as in the console (at least using monospaced fonts).

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 11]
```diff
@@ -0,0 +1,162 @@
+﻿using NUnit.Framework;
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Missing copyright notice

File: src/tests/Model/ResultSummaryReporterTests.cs [hunk 12]
```diff
@@ -0,0 +1,135 @@
+﻿using System;
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryReporterTests.cs]: Missing copyright notice

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+namespace NUnit.Gui.Model
+{
+    [TestFixture]
+    public class ResultSummaryCreatorTests
+    {
+        [Test]
+        public void WhenResultNodeIsNotForTestRunExceptionIsThrown()
+        {
+            Assert.That(() => CreateResultSummary("<anything-other-than-test-run/>"), Throws.InstanceOf<InvalidOperationException>());
+
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Insert newline before `Throws.InstanceOf....`

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.That(() => CreateResultSummary("<test-run/>"), Throws.Nothing);
+        }
+        
+        [Test]
+        public void WhenResultIsNotSpecified_PassedIsDefault()
+        {
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Perhaps write this as two separate tests?

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.That(summary.OverallResult, Is.EqualTo("Passed"));
+
+            summary = CreateResultSummary("<test-run result='Failed'/>");
+
+            Assert.That(summary.OverallResult, Is.EqualTo("Failed"));
+        }
+
+        [Test]
+        public void WhenDurationIsNotSpecified_ZeroIsDefault()
+        {
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Perhaps write this as two separate tests? Then we can also use a TestCase for duration to test other durations.

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.That(summary.Duration, Is.EqualTo(0.0));
+
+            summary = CreateResultSummary("<test-run duration='1.9'/>");
+
+            Assert.That(summary.Duration, Is.EqualTo(1.9));
+        }
+
+        [Test]
+        public void WhenStartTimeIsNotSpecified_DateTimeMinValueIsDefault()
+        {
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Perhaps write this as two separate tests? Then we can also use a TestCase for StartTime to test other starttimes.

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+                    "<test-case/>" +
+                    "<test-case/>" +
+                "</test-suite>";
+            var summary = CreateResultSummary($"<test-run>{innerXml}</test-run>");
+
+            Assert.That(summary.TestCount, Is.EqualTo(4));
+        }
+
+        [Test]
+        public void WhenNoResultIsNotSpecifiedInTestCase_SkipCountIsIncremented()
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Remove Not from test name. E.g.
WhenNoResultIsNotSpecifiedInTestCase_SkipCountIsIncremented => WhenNoResultIsSpecifiedInTestCase_SkipCountIsIncremented

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.That(summary.UnexpectedError, Is.False);
+        }
+
+        [Test]
+        public void ErrorAssembliesMarkSummaryAsUnexpectedError()
+        {
+            var innerXml =
+                "<test-suite result='Failed' label='Error' type='Assembly'/>" +
+                "<test-suite result='Failed' label='Error' type='Assembly'/>";
+            var summary = CreateResultSummary($"<test-run>{innerXml}</test-run>");
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Perhaps change the label of the first test-suite to "Invalid"?

File: src/tests/Model/ResultSummaryCreatorTests.cs [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+            var expectedDate = new DateTime(2017, 7, 8, 6, 19, 23);
+            summary = CreateResultSummary($"<test-run start-time='{expectedDate.ToString("u")}'/>");
+
+            Assert.That(summary.StartTime, Is.EqualTo(expectedDate));
+        }
+
+        [Test]
+        public void WhenEndTimeIsNotSpecified_DateTimeMaxValueIsDefault()
+        {
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: Perhaps write this as two separate tests? Then we can also use a TestCase for EndTime to test other endtimes.

File: src/tests/Model/ResultSummaryReporterTests.cs [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace NUnit.Gui.Model
+{
+    [TestFixture]
+    public class ResultSummaryReporterTests
+    {
+        [Test]
+        public void ReportAlwaysStartsWithOveralResult()
+        {
+            var summary = new ResultSummary{ OverallResult = "Passed"};
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryReporterTests.cs]: Missing spaces. `ResultSummary{ OverallResult = "Passed"}` => `ResultSummary { OverallResult = "Passed" }`

File: src/tests/Model/ResultSummaryReporterTests.cs [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+                FailureCount = failureCount,
+                ErrorCount = errorCount,
+                InvalidCount = invalidCount,
+                OverallResult = String.Empty
+            };
+
+            var report = ResultSummaryReporter.WriteSummaryReport(summary);
+            Assert.That(report, Does.Contain("Failed Tests - Failures:"));
+            Assert.That(report, Does.Contain("Errors:"));
+            Assert.That(report, Does.Contain("Invalid:"));
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryReporterTests.cs]: I think that we should assert that the correct numbers are presented in the report.

File: src/tests/Model/ResultSummaryReporterTests.cs [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+                ExplicitCount = explicitCount,
+                SkipCount = skipCount,
+                OverallResult = String.Empty
+            };
+
+            var report = ResultSummaryReporter.WriteSummaryReport(summary);
+
+            Assert.That(report, Does.Contain("Skipped Tests - Ignored:"));
+            Assert.That(report, Does.Contain("Explicit:"));
+            Assert.That(report, Does.Contain("Other:"));
```
Comment (mikkelbu) [inline src/tests/Model/ResultSummaryReporterTests.cs]: I think that we should assert that the correct counts are presented in the report.

File: src/tests/Presenters/StatusBarPresenterTests.cs [hunk 23]
```diff
@@ -95,6 +95,17 @@ public void WhenTestCaseCompletes_CountIsIncremented()
             _view.Received().OnTestPassed();
         }
 
+        [Test]
+        public void WhenTestsFinish_TestResultsAreCalculated()
+        {
+            var result = new ResultNode(XmlHelper.CreateXmlNode("<test-run/>"));
+
+            _model.RunFinished += Raise.Event<TestResultEventHandler>(new TestResultEventArgs(TestAction.TestFinished, result));
```
Comment (mikkelbu) [inline src/tests/Presenters/StatusBarPresenterTests.cs]: `TestAction.TestFinished` => `TestAction.RunFinished` ?

**Comment by mikkelbu** [comment]
With respect to the "anchoring" I mean that the first time I run tests it looks like this 
image
and if I then rerun the tests, then it looks like this 
image

File: src/nunit-gui/Model/ResultSummary.cs (hunk 6 above)
Comment (immeraufdemhund) [inline src/nunit-gui/Model/ResultSummary.cs]: this model was taken from the nunit-console project, and the calculation of the members is all from their too. Because of that I'm not sure how I could/would rephrase this. The only thing I do know is that if the test case's result is failure and label is "Invalid" then this value is increased. How or why the value is returned from the test-engine is beyond me.

File: src/nunit-gui/Model/ResultSummaryReporter.cs (hunk 8 above)
Comment (immeraufdemhund) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: To be honest I was wondering about the same thing my self. I decided in the end to keep the string writer and color around because of how it is done in the console. I've seen a few examples of where rich text was allowed in the tooltip and if we decide to make a WPF implementation then it would be nice to copy...

then again it violates YAGNI...*sigh* ok, i'll simplify.

File: src/tests/Model/ResultSummaryCreatorTests.cs (hunk 18 above)
Comment (immeraufdemhund) [inline src/tests/Model/ResultSummaryCreatorTests.cs]: good catch, copy and paste error on my part

**Comment by immeraufdemhund** [comment]
I see what you are saying about the ancoring. It's doing it on my PC. I'm not sure what or why this is happening so I'll leave it for tonight and see if I can figure it out tomorrow.

File: src/nunit-gui/Model/ResultSummary.cs (hunk 6 above)
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummary.cs]: This is a very minor detail, so we can keep the comment the same as in the nunit-console project. I just found it strange, when we also have `public int FailedCount => FailureCount + InvalidCount + ErrorCount;` in the same file. So as far as I can tell `InvalidCount` are also counted as failed, not as erros.

File: src/nunit-gui/Model/ResultSummaryReporter.cs [hunk 24]
```diff
@@ -31,57 +30,43 @@ public static class ResultSummaryReporter
...
-        }
 
-        private static void WriteLabelLine(this TextWriter writer, string line, object option)
-        {
-            writer.WriteLabel(line, option);
-            writer.WriteLine();
-        }
+            writer.AppendLine($"  Start time: {summary.StartTime:u}");
+            writer.AppendLine($"    End time: {summary.EndTime:u}");
+            writer.AppendLine($"    Duration: {summary.Duration:0.000}");//, string.Format(NumberFormatInfo.InvariantInfo, "{0:0.000} seconds", summary.Duration
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: Please remove the comment

File: src/nunit-gui/Model/ResultSummaryReporter.cs [hunk 25]
```diff
@@ -31,57 +30,43 @@ public static class ResultSummaryReporter
...
 
-                writer.WriteLabelLine("  Start time: ", summary.StartTime.ToString("u"));
-                writer.WriteLabelLine("    End time: ", summary.EndTime.ToString("u"));
-                writer.WriteLabelLine("    Duration: ", string.Format(NumberFormatInfo.InvariantInfo, "{0:0.000} seconds", summary.Duration));
+            writer.AppendUICultureFormattedNumber($"  Test Count: ", summary.TestCount);
+            writer.AppendUICultureFormattedNumber($", Passed: ", summary.PassCount);
+            writer.AppendUICultureFormattedNumber($", Failed: ", summary.FailedCount);
+            writer.AppendUICultureFormattedNumber($", Warnings: ", summary.WarningCount);
+            writer.AppendUICultureFormattedNumber($", Inconclusive: ", summary.InconclusiveCount);
+            writer.AppendUICultureFormattedNumber($", Skipped: ", summary.TotalSkipCount);
```
Comment (mikkelbu) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: Why all the $"...." when we are not using any parameters. For clarity I think that we should remove those $ from line 36-56.

Comment (immeraufdemhund) [inline src/nunit-gui/Model/ResultSummaryReporter.cs]: oops, remnants of refactoring

**Comment by immeraufdemhund** [comment]
I don't personally like to have multiple commits of "fixing this and that" would you mind if I squash my two commits after you've checked them out? Also I haven't figured out the tool tip thing.

**Comment by mikkelbu** [comment]
Please squash :)

Regarding the tooltip I cannot see anything wrong in your code (note that I'm far from a gui-expert), and personally I don't think it is a major issue. Personally, I think that it is okay to merge this branch to master, and then we can see if we can solve the anchoring in a separate issue.

**Comment by immeraufdemhund** [comment]
ugh.. it messed up the commit message *sigh*
